src/sources/soundsourceffmpeg.cpp

```diff
@@ -0,0 +1,533 @@
+                            m_bUnique = true;
+                            l_lLastPacketPos = l_SPacket.pos;
+                        }
+
+                        // If we are over last storepos and we have read more than jump point need is and pos is unique we store it to memory
+                        if (m_lCacheFramePos > m_lLastStoredPos &&
+                                m_lCacheFramePos > (AUDIOSOURCEFFMPEG_POSDISTANCE + m_lLastStoredPos) &&
+                                m_bUnique == true) {
+                            struct ffmpegLocationObject  *l_SJmp = (struct ffmpegLocationObject  *)malloc(
+                                    sizeof(struct ffmpegLocationObject));
+                            m_lLastStoredPos = m_lCacheFramePos;
+                            l_SJmp->startFrame = m_lCacheFramePos;
+                            l_SJmp->pos = l_SPacket.pos;
+                            l_SJmp->pts = l_SPacket.pts;
+                            m_SJumpPoints.append(l_SJmp);
+                            m_bUnique = false;
+                        }
+
+                        if (offset < 0 || offset <= m_lCacheFramePos) {
+                            l_iCount --;
+                        }
+                    } else {
+                        free(l_SObj);
+                        l_SObj = NULL;
+                        qDebug() <<
+                                 "SoundSourceFFmpeg::readFramesToCache: General error in audio decode:" <<
+                                 l_iRet;
+                    }
+                }
+
+                av_free_packet(&l_SPacket);
+                l_SPacket.data = NULL;
+                l_SPacket.size = 0;
+
+            } else {
+                l_iError ++;
+                if (l_iError == 5) {
+                    // Stream end and we couldn't read enough frames
+                    l_iStop = true;
+                }
+            }
+
+
+        } else {
+            qDebug() << "SoundSourceFFmpeg::readFramesToCache: Packet too big or File end";
+            l_iStop = true;
+        }
+
+    }
+
+    if (l_pFrame != NULL) {
+      l_iFrameCount --;
+// FFMPEG 2.2 3561060 anb beyond
+#if LIBAVCODEC_VERSION_INT >= 3561060
+          av_frame_unref(l_pFrame);
+          av_frame_free(&l_pFrame);
+// FFMPEG 0.11 and below
+#elif LIBAVCODEC_VERSION_INT <= 3544932
+          av_free(l_pFrame);
+// FFMPEG 1.0 - 2.1
+#else
+          avcodec_free_frame(&l_pFrame);
+#endif
+          l_pFrame = NULL;
+
+    }
+
+    if (l_iFrameCount > 0) {
+       qDebug() << "SoundSourceFFmpeg::readFramesToCache(): Frame balance is not 0 it is: " << l_iFrameCount;
+    }
+
+    l_SObj = m_SCache.first();
+    m_lCacheStartFrame = l_SObj->startFrame;
+    l_SObj = m_SCache.last();
+    m_lCacheEndFrame = (l_SObj->startFrame + l_SObj->length);
+
+    if (!l_iCount) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+bool SoundSourceFFmpeg::getBytesFromCache(char *buffer, SINT offset,
```

Comment:
@illuusio:
This function is somehow misleading since size is the number of Fames but the buffer destination buffer is of type CSAMPLE*

It is probably good to use getSamplesFromCache and move the channel multiplier to the calling code. 

Nitpick: 
It would also be great if you could use a const for the channel count 2 in various places. 
Uwe uses conversion helpers. Maybe you can reuse them. 

Keep in mind that it is required to make the channel count changeable when we try to "fix" this bug: 
https://bugs.launchpad.net/mixxx/+bug/1438746